metrics.go

```diff
@@ -0,0 +1,416 @@
+	}
+}
+
+func Hour() MetricUnit {
+	return MetricUnit{
+		"hour",
+	}
+}
+
+func Day() MetricUnit {
+	return MetricUnit{
+		"day",
+	}
+}
+
+func Week() MetricUnit {
+	return MetricUnit{
+		"week",
+	}
+}
+
+func Bit() MetricUnit {
+	return MetricUnit{
+		"bit",
+	}
+}
+
+func Byte() MetricUnit {
+	return MetricUnit{
+		"byte",
+	}
+}
+
+func KiloByte() MetricUnit {
+	return MetricUnit{
+		"kilobyte",
+	}
+}
+
+func KibiByte() MetricUnit {
+	return MetricUnit{
+		"kibibyte",
+	}
+}
+
+func MegaByte() MetricUnit {
+	return MetricUnit{
+		"megabyte",
+	}
+}
+
+func MebiByte() MetricUnit {
+	return MetricUnit{
+		"mebibyte",
+	}
+}
+
+func GigaByte() MetricUnit {
+	return MetricUnit{
+		"gigabyte",
+	}
+}
+
+func GibiByte() MetricUnit {
+	return MetricUnit{
+		"gibibyte",
+	}
+}
+
+func TeraByte() MetricUnit {
+	return MetricUnit{
+		"terabyte",
+	}
+}
+
+func TebiByte() MetricUnit {
+	return MetricUnit{
+		"tebibyte",
+	}
+}
+
+func PetaByte() MetricUnit {
+	return MetricUnit{
+		"petabyte",
+	}
+}
+
+func PebiByte() MetricUnit {
+	return MetricUnit{
+		"pebibyte",
+	}
+}
+
+func ExaByte() MetricUnit {
+	return MetricUnit{
+		"exabyte",
+	}
+}
+
+func ExbiByte() MetricUnit {
+	return MetricUnit{
+		"exbibyte",
+	}
+}
+
+func Ratio() MetricUnit {
+	return MetricUnit{
+		"ratio",
+	}
+}
+
+func Percent() MetricUnit {
+	return MetricUnit{
+		"percent",
+	}
+}
+
+func CustomUnit(unit string) MetricUnit {
+	re := regexp.MustCompile(`[^a-z]+`)
+	return MetricUnit{
+		re.ReplaceAllString(unit, ""),
+	}
+}
+
+type Metric interface {
+	GetType() string
+	GetTags() map[string]string
+	GetKey() string
+	GetUnit() string
+	GetTimestamp() int
+	SerializeValue() string
+	SerializeTags() string
+}
+
+type abstractMetric struct {
+	key       string
+	unit      MetricUnit
+	tags      map[string]string
+	timestamp int
+}
+
+func (am abstractMetric) GetTags() map[string]string {
+	return am.tags
+}
+
+func (am abstractMetric) GetKey() string {
+	return am.key
+}
+
+func (am abstractMetric) GetUnit() string {
+	return am.unit.toString()
+}
+
+func (am abstractMetric) GetTimestamp() int {
+	return am.timestamp
+}
+
+func (am abstractMetric) SerializeTags() string {
+	var sb strings.Builder
+
+	values := make([]string, 0, len(am.tags))
+	for k := range am.tags {
+		values = append(values, k)
+	}
+	sortSlice(values)
+
+	for _, key := range values {
+		val := sanitizeValue(am.tags[key])
+		key = sanitizeKey(key)
+		sb.WriteString(fmt.Sprintf("%s:%s,", key, val))
+	}
+	s := sb.String()
+	if len(s) > 0 {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+// Counter Metric.
+type CounterMetric struct {
+	value float64
+	abstractMetric
+}
+
+func (c *CounterMetric) Add(value float64) {
+	c.value += value
+}
+
+func (c CounterMetric) GetType() string {
+	return "c"
+}
+
+func (c CounterMetric) SerializeValue() string {
+	return fmt.Sprintf(":%v", c.value)
+}
+
+func NewCounterMetric(key string, unit MetricUnit, tags map[string]string, timestamp int, value float64) CounterMetric {
```

Comment:
I'd prefer `timestamp` to be `time.Time` instead of `int` or any derivative of integers. The SDK should serialize it, instead of throwing it to users.

Reply:
This is a fair suggestion, but this structs and functions that we're adding in this PR are just the foundation for metrics support and won't be used by users mostly. 

Users will use the high level APIs that will be added later in a different PR. So that's where we could accept `Time` and make the appropriate conversion.